academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1623 @@
+
+    def download_baseline(self, release: PubMedRelease, **kwargs):
+        """
+        Download files from PubMed's FTP server for this release.
+
+        Unable to do this in parallel because their FTP server is not able to handle too many requests.
+        """
+
+        if not release.year_first_run:
+            logging.info("download_baseline: skipping as this is only run when there is a new baseline dump available.")
+            return
+
+        baseline_datafiles = [datafile for datafile in release.datafile_list if datafile.baseline]
+
+        success = download_datafiles_from_ftp_server(
+            datafile_list=baseline_datafiles,
+            ftp_server_url=self.ftp_server_url,
+            ftp_port=self.ftp_port,
+            reset_ftp_counter=self.reset_ftp_counter,
+            max_download_retry=self.max_download_retry,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def upload_downloaded_baseline(self, release: PubMedRelease, **kwargs):
+        """Upload downloaded baseline files to GCS."""
+
+        if not release.year_first_run:
+            logging.info(
+                "upload_downloaded_baseline: skipping as this is only run when there is a new baseline dump available."
+            )
+            return
+
+        # Grab list of files to upload.
+        datafiles_to_upload = [datafile.download_file_path for datafile in release.datafile_list if datafile.baseline]
+
+        success = gcs_upload_files(
+            bucket_name=self.cloud_workspace.download_bucket,
+            file_paths=datafiles_to_upload,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def transform_baseline(self, release: PubMedRelease, **kwargs):
+        """
+        Transform the *.xml.gz files downloaded from PubMed into usable json files for BigQuery import.
+
+        This is a multithreaded task that pulls the information from the files using the keys defined in self.entity_list.
+        """
+
+        if not release.year_first_run:
+            logging.info(
+                "transform_baseline: skipping as this is only run when there is a new baseline dump available."
+            )
+            return
+
+        # Get list of datafiles for the baseline to trasform.
+        datafiles_to_transform = [datafile for datafile in release.datafile_list if datafile.baseline]
+
+        # Process files in batches so that ProcessPoolExecutor doesn't deplete the system of memory
+        for i, chunk in enumerate(get_chunks(input_list=datafiles_to_transform, chunk_size=self.max_processes)):
+            with ProcessPoolExecutor(max_workers=self.max_processes) as executor:
+                logging.info(f"In chunk {i} and processing files: {chunk}")
+
+                futures = [
+                    executor.submit(transform_pubmed_xml_file_to_jsonl, datafile, [self.entity_list["baseline"]])
+                    for datafile in chunk
+                ]
```

Comment:
I think it would be clearer to explicitly pass in what files to transform (see suggestions at the end on the transforming code):

```python
futures = []
for datafile in chunk:
    input_path = datafile.download_file_path
    upsert_path = datafile.transform_file_path(entity_type=self.baseline.type)
    futures.append(executor.submit(transform_pubmed, input_path, upsert_path))
```